### PR TITLE
Allow MAVExplorer Tools to be used in Windows

### PR DESCRIPTION
**Pull request by stephendade**
This patch allows the programs in the "Tools" menu of MAVExplorer to be used in Windows.

Previously, MAVExplorer would crash when starting any of these tools (see https://discuss.ardupilot.org/t/magfit-mavexplorer-on-windows/75138).

The root cause of the issue is that Python is unable to pickle a mlog object when passing the mlog to a new process (in Windows. Linux is fine).

I was unable to find a method to make mlog pickle-able, so decided to go with making all Tools use threads instead of processes.

### CODE REVIEW

File: MAVProxy/modules/lib/multiproc_util.py [hunk 1]
```diff
@@ -10,6 +10,8 @@
 import mmap
 import platform
 import struct
+import threading
+import os
```
Comment (peterbarker) [inline MAVProxy/modules/lib/multiproc_util.py:14]: Put in alphabetical order

File: MAVProxy/modules/lib/multiproc_util.py [hunk 2]
```diff
@@ -198,10 +200,16 @@ def start(self):
         with mutex:
             self.wrap()
             try:
-                self._child = multiproc.Process(target=self._child_task)
+                if os.name == 'nt':
+                    # Windows can't pickle the mavlog object, thus can't spin off
+                    #
```
Comment (peterbarker) [inline MAVProxy/modules/lib/multiproc_util.py:205]:
```suggestion
```

File: MAVProxy/modules/lib/wxconsole_ui.py [hunk 3]
```diff
@@ -93,52 +93,52 @@ def on_timer(self, event):
...
+                    if obj is not None:
+                        self.SetMenuBar(None)
+                        self.menu = obj
+                        self.SetMenuBar(self.menu.wx_menu())
+                        self.Bind(wx.EVT_MENU, self.on_menu)
+                    self.Refresh()
+                    self.Update()
+                elif isinstance(obj, win_layout.WinLayout):
+                    win_layout.set_wx_window_layout(self, obj)
+        except (EOFError, BrokenPipeError):
```
Comment (peterbarker) [inline MAVProxy/modules/lib/wxconsole_ui.py:141]:
```suggestion
        except EOFError, BrokenPipeError:
```

File: MAVProxy/modules/lib/wxconsole_ui.py [hunk 4]
```diff
@@ -93,52 +93,52 @@ def on_timer(self, event):
             self.timer.Stop()
             self.Destroy()
             return
-        while state.child_pipe_recv.poll():
-            try:
+        try:
```
Comment (peterbarker) [inline MAVProxy/modules/lib/wxconsole_ui.py:96]: Why are you shifting the `try` here?

It increases the code covered by the try/except block - when the exception being caught should really only come from the line that's currently covered by the try/except block, shouldn't it?

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: MAVProxy/modules/lib/wxconsole_ui.py (hunk 4 above)
Comment (stephendade) [reply MAVProxy/modules/lib/wxconsole_ui.py:96]: Similar to https://github.com/ArduPilot/MAVProxy/commit/4898e10555f22a0db5a3e36be7243b568ca6c732, the ``BrokenPipeException`` occurs when ``poll()`` is called. Not sure if there's a cleaner way to do this...

**Comment by stephendade** [comment]
Updated as per you comments

File: MAVProxy/modules/lib/wxconsole_ui.py [hunk 5]
```diff
@@ -93,52 +93,52 @@ def on_timer(self, event):
...
+                    if obj is not None:
+                        self.SetMenuBar(None)
+                        self.menu = obj
+                        self.SetMenuBar(self.menu.wx_menu())
+                        self.Bind(wx.EVT_MENU, self.on_menu)
+                    self.Refresh()
+                    self.Update()
+                elif isinstance(obj, win_layout.WinLayout):
+                    win_layout.set_wx_window_layout(self, obj)
+        except EOFError, BrokenPipeError:
```
Comment (peterbarker) [inline MAVProxy/modules/lib/wxconsole_ui.py:141]: See https://github.com/ArduPilot/MAVProxy/pull/946 - `BrokenPipeError` doesn't exist under Python2

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: MAVProxy/modules/lib/wxconsole_ui.py (hunk 5 above)
Comment (stephendade) [reply MAVProxy/modules/lib/wxconsole_ui.py]: Cool - i've incorporated that change now

**Pull request merged**